djstripe/stripe_objects.py

```diff
@@ -0,0 +1,529 @@
+
+    Temporarily replace the global api_key used in stripe API calls with the given value
+    the original value is restored as soon as context exits
+    """
+    import stripe
+    bkp_key = stripe.api_key
+    stripe.api_key = temp_key
+    yield
+    stripe.api_key = bkp_key
+
+
+def convert_tstamp(response, field_name=None):
+    # Overrides the set timezone to UTC - I think...
+    tz = timezone.utc if settings.USE_TZ else None
+
+    if not field_name:
+        return datetime.datetime.fromtimestamp(response, tz)
+    else:
+        if field_name in response and response[field_name]:
+            return datetime.datetime.fromtimestamp(response[field_name], tz)
+
+
+@python_2_unicode_compatible
+class StripeObject(TimeStampedModel):
+    # This must be defined in descendants of this model/mixin
+    # e.g. "Event", "Charge", "Customer", etc.
+    stripe_api_name = None
+    stripe_objects = StripeObjectManager()
+
+    stripe_id = models.CharField(max_length=50, unique=True)
+    # livemode = models.BooleanField(default=False)
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def api(cls):
+        """
+        Get the api object for this type of stripe object (requires
+        stripe_api_name attribute to be set on model).
+        """
+        if cls.stripe_api_name is None:
+            raise NotImplementedError("StripeObject descendants are required to define "
+                                      "the stripe_api_name attribute")
+        # e.g. stripe.Event, stripe.Charge, etc
+        return getattr(stripe, cls.stripe_api_name)
+
+    def api_retrieve(self):
+        """
+        Implement very commonly used API function 'retrieve'
+        """
+        # Run stripe.X.retreive(id)
+        return type(self).api().retrieve(self.stripe_id)
+
+    @classmethod
+    def api_create(cls, **kwargs):
+        """
+        Call the stripe API's create operation for this model
+        """
+        return cls.api().create(**kwargs)
+
+    def str_parts(self):
+        """
+        extend this to add information to the string representation of the object
+        :rtype: list of str
+        """
+        return ["stripe_id={id}".format(id=self.stripe_id)]
+
+    @classmethod
+    def stripe_obj_to_record(cls, data):
+        raise NotImplementedError()
+
+    @classmethod
+    def create_from_stripe_object(cls, data):
+        """
+        Create a model instance (not saved to db), using the given data object from Stripe
+        :type data: dict
+        """
+        return cls(**cls.stripe_obj_to_record(data))
+
+    def __str__(self):
+        return "<{list}>".format(list=", ".join(self.str_parts()))
+
+
+class StripeEvent(StripeObject):
+    stripe_api_name = "Event"
+
+    class Meta:
+        abstract = True
+    livemode = models.BooleanField(default=False)
+
+    # This is "type" in Stripe
+    kind = models.CharField(max_length=250)
+    # This is "data" in Stripe
+    webhook_message = JSONField()
```

Comment:
Since we're modifying our API heavily right now anyway, feel free to rename these and any other fields that don't match up correctly.